wincode/src/io/vec.rs

```diff
@@ -0,0 +1,164 @@
+use super::*;
+
+/// Trusted writer for `Vec<u8>` that continues appending to the vector's spare capacity.
+///
+/// This will _not_ grow the vector, as it assumes the caller has already reserved enough capacity.
+pub struct TrustedVecWriter<'a> {
+    inner: &'a mut Vec<u8>,
+}
+
+impl<'a> TrustedVecWriter<'a> {
+    pub fn new(inner: &'a mut Vec<u8>) -> Self {
+        Self { inner }
+    }
+}
+
+impl<'a> Writer for TrustedVecWriter<'a> {
+    type Trusted<'b>
+        = TrustedVecWriter<'b>
+    where
+        Self: 'b;
+
+    fn write(&mut self, src: &[u8]) -> WriteResult<()> {
+        let spare = self.inner.spare_capacity_mut();
+        // SAFETY: Creator of this writer ensures we have sufficient capacity for all writes.
+        unsafe { ptr::copy_nonoverlapping(src.as_ptr(), spare.as_mut_ptr().cast(), src.len()) };
```

Comment:
This will cause UB if the spare capacity is smaller than `src`. I know that the point of `TrustedVecWriter` is to never grow, but I still think we could return an error, especially since `WriteError::WriteSizeLimit` looks like a great fit.

Or if we really want to shift the responsibility on the caller, this method needs to be `unsafe`, but that defeats the usage of the trait.

Reply:
Fair. But that expectation means this method needs to be `unsafe` (it can cause UB if misused), especially given that it's public. By making the method public and safe we are not making the expectation clear. And it's not even 100% clear from reading any docstring (the second paragraph of the `TrustedVector`'s description mentions capacity, but doesn't say that `write` is unchecked and can cause UB), so the caller needs to explicitly read the code to understand the risk. :upside_down_face: 

I don't think that adding a check with an error will kill the performance. And adding a separate trait with unsafe methods for writing seems like unnecessary work (I would be happy with it though if you think otherwise).